modules/api/src/main/java/org/apache/ignite/lang/ErrorGroups.java

```diff
@@ -354,6 +354,9 @@ public static class Transactions {
 
         /** Coordinator tries to commit a transaction that has already been aborted. */
         public static final int TX_WAS_ABORTED_ERR = TX_ERR_GROUP.registerErrorCode((short) 15);
+
+        /** Coordinator tries to abort a transaction that has already been committed. */
+        public static final int TX_WAS_COMMITTED_ERR = TX_ERR_GROUP.registerErrorCode((short) 16);
```

Comment:
You should also check this code in `TxManagerImpl#durableFinish` to stop durable finish. By the way, seems we dont return the correct error code from `PartitionListener#onTxStateStorageCasFail` - there is `TX_UNEXPECTED_STATE_ERR` in any case, should be either `TX_WAS_ABORTED_ERR` or `TX_WAS_COMMITTED_ERR`.

Reply:
I am not sure, maybe it still can be useful for some cases of parallel recovery to see the actual state in FSM